core/src/component/registry.rs

```diff
@@ -208,4 +203,36 @@ where
             .get_mut(index)
             .and_then(|box_component| (*(*box_component)).as_mut_any().downcast_mut())
     }
+
+    /// Register an individual component.
+    ///
+    /// This is an internal method used by `Registry::register`.
+    /// It shouldn't be exposed through the public API without careful
+    /// consideration, as it's not yet designed to be used outside of
+    /// that particular context.
+    fn register_component(
+        &mut self,
+        component: Box<dyn Component<A>>,
+    ) -> Result<(), FrameworkError> {
+        let id = component.id();
+        let version = component.version();
+        let type_id = (*component).type_id();
+        let index = self.components.insert(component);
+
+        // Index component by ID
+        if self.id_map.insert(id, index).is_some() {
+            self.components.remove(index);
```

Comment:
Is this correct? Quick scan, may be wrong, but looks like you are removing the original component when you try and add a duplicate?

Reply:
Looking at this, it is wrong, but not for the reason you suspect. It's possible for two instances of the same component to coexist in the `Arena`, and that's the `index` you're seeing (i.e. the second copy).

However, this is removing the original index from the map when it encounters a duplicate ID, so the map becomes inconsistent, which is a bug.

I think a better approach is to test both maps using `contains_key` prior to doing any state mutations, rather than getting into an inconsistent state in the first place.

Thanks for pointing it out!

Reply:
PR to fix the buggy registration here: #122